src/ess/reduce/time_of_flight/toa_to_tof.py

```diff
@@ -0,0 +1,451 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2025 Scipp contributors (https://github.com/scipp)
+"""
+Time-of-flight workflow for unwrapping the time of arrival of the neutron at the
+detector.
+This workflow is used to convert raw detector data with event_time_zero and
+event_time_offset coordinates to data with a time-of-flight coordinate.
+"""
+
+from collections.abc import Callable
+from functools import reduce
+
+import numpy as np
+import scipp as sc
+from scipp._scipp.core import _bins_no_validate
+from scippneutron._utils import elem_unit
+
+from ..nexus.types import DetectorData, RunType
+from .to_events import to_events
+from .types import (
+    DistanceResolution,
+    Facility,
+    FrameFoldedTimeOfArrival,
+    FramePeriod,
+    LookupTableVarianceThreshold,
+    Ltotal,
+    MaskedTimeOfFlightLookupTable,
+    NumberOfNeutrons,
+    PivotTimeAtDetector,
+    PulsePeriod,
+    PulseStride,
+    PulseStrideOffset,
+    ReHistogrammedTofData,
+    SimulationResults,
+    SimulationSeed,
+    TimeOfArrivalMinusPivotTimeModuloPeriod,
+    TimeOfFlightLookupTable,
+    TofData,
+    UnwrappedTimeOfArrival,
+    UnwrappedTimeOfArrivalMinusPivotTime,
+)
+
+
+def pulse_period_from_source(facility: Facility) -> PulsePeriod:
+    """
+    Return the period of the source pulses, i.e., time between consecutive pulse starts.
+
+    Parameters
+    ----------
+    facility:
+        Facility where the experiment is performed (used to determine the source pulse
+        parameters).
+    """
+    facilities = {"ess": sc.scalar(14.0, unit="Hz")}
+    return PulsePeriod(1.0 / facilities[facility])
+
+
+def frame_period(pulse_period: PulsePeriod, pulse_stride: PulseStride) -> FramePeriod:
+    """
+    Return the period of a frame, which is defined by the pulse period times the pulse
+    stride.
+
+    Parameters
+    ----------
+    pulse_period:
+        Period of the source pulses, i.e., time between consecutive pulse starts.
+    pulse_stride:
+        Stride of used pulses. Usually 1, but may be a small integer when
+        pulse-skipping.
+    """
+    return FramePeriod(pulse_period * pulse_stride)
+
+
+def extract_ltotal(da: DetectorData[RunType]) -> Ltotal[RunType]:
+    """
+    Extract the total length of the flight path from the source to the detector from the
+    detector data.
+
+    Parameters
+    ----------
+    da:
+        Raw detector data loaded from a NeXus file, e.g., NXdetector containing
+        NXevent_data.
+    """
+    return Ltotal[RunType](da.coords["Ltotal"])
```

Comment:
Similarly to the source, does it make sense to have a Beamline class / module to encode common data? I am also thinking of metadata like the beamline's name, facility, etc, which would be useful for writing output files. See also https://github.com/scipp/scippneutron/issues/473